javaparser-core-testing/src/test/java/com/github/javaparser/ast/expr/InstanceOfExprTest.java

```diff
@@ -21,16 +21,97 @@
+import static com.github.javaparser.ParserConfiguration.LanguageLevel;
 import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
 
+/**
+ * See JEP305 for switch pattern matching for instanceof
+ * https://bugs.openjdk.java.net/browse/JDK-8181287
+ *
+ * <pre>
+ * The instanceof grammar is extended accordingly:
+ *
+ * RelationalExpression:
+ *      ...
+ *      RelationalExpression instanceof ReferenceType
+ *      RelationalExpression instanceof Pattern
+ *
+ * Pattern:
+ *      ReferenceType Identifier
+ * </pre>
+ */
 class InstanceOfExprTest {
+
+    @Test
+    void annotationsOnTheType_patternExpression() {
+        InstanceOfExpr expr = StaticJavaParser.parseExpression("obj instanceof @A @DA String s");
+
+        assertEquals("obj", expr.getExpression().toString());
+        assertEquals("String", expr.getType().asString());
+        assertTrue(expr.getName().isPresent());
+
+        assertThat(expr.getType().getAnnotations())
+                .containsExactly(
+                        new MarkerAnnotationExpr("A"),
+                        new MarkerAnnotationExpr("DA")
+                );
+    }
+
     @Test
-    void annotationsOnTheType() {
-        InstanceOfExpr expr = parseExpression("s instanceof @A @DA String");
+    void annotationsOnTheType_referenceTypeExpression() {
+        InstanceOfExpr expr = StaticJavaParser.parseExpression("obj instanceof @A @DA String");
+
+        assertEquals("obj", expr.getExpression().toString());
+        assertEquals("String", expr.getType().asString());
+        assertFalse(expr.getName().isPresent());
 
         assertThat(expr.getType().getAnnotations()).containsExactly(new MarkerAnnotationExpr("A"), new MarkerAnnotationExpr("DA"));
     }
+
+    @Test
+    void instanceOf_patternExpression() {
+        String x = "obj instanceof String s";
+        InstanceOfExpr expr = TestParser.parseExpression(LanguageLevel.JAVA_14, x);
+
+        assertEquals("obj", expr.getExpression().toString());
+        assertEquals("String", expr.getType().asString());
+        assertTrue(expr.getName().isPresent());
+    }
+
+    @Test
+    void instanceOf_referenceTypeExpression() {
+        String x = "obj instanceof String";
+        InstanceOfExpr expr = TestParser.parseExpression(LanguageLevel.JAVA_14, x);
+
+        assertEquals("obj", expr.getExpression().toString());
+        assertEquals(String.class.getSimpleName(), expr.getType().asString());
+        assertFalse(expr.getName().isPresent());
+    }
+
+
+
+    /*
+     * resolution / scoping tests?
+     *
+     * <pre>
+     * {@code
+     * if (!(obj instanceof String s)) {
+     *     .. s.contains(..) ..
+     * } else {
+     *     .. s.contains(..) ..
+     * }
+     * }
+     * </pre>
+     *
+     * Allowed / in scope: {@code if (obj instanceof String s && s.length() > 5) {..}}
+     * Not in scope:       {@code if (obj instanceof String s || s.length() > 5) {..}}
+     *
```

Comment:
I wouldn't add validations for that, I bet they're going to get really complex. Avoid building a compiler :-)

Reply:
It will become relevant in the solver part I think, trying to resolve s.length()

But this is perhaps a compilation-time error..

At the very least we should include this as a note in the docs re: known limitations :)

Reply:
Ah yes, to keep from going crazy I always tried to delegate the symbol solver changes to someone else :-#